File: server/dsp/process.py
"""
process.py — Real DSP processing pipeline using Pedalboard
Applies: EQ (HighpassFilter, LowpassFilter, PeakFilter bands),
         Compressor, Reverb, Limiter, Stereo Width
"""
import json
import sys
import numpy as np
import soundfile as sf
import pyloudnorm as pyln
from pedalboard import (
    Pedalboard,
    HighpassFilter,
    LowpassFilter,
    PeakFilter,
    Compressor,
    Reverb,
    LowShelfFilter,
    HighShelfFilter,
    Limiter,
    Gain,
)


def apply_dsp(input_path: str, output_path: str, settings: dict) -> dict:
    """
    Apply the full DSP chain to input_path and write to output_path (WAV).
    settings: dict with keys eq, compression, reverb, stereo, lufs_target
    Returns a summary dict.
    """
    data, sr = sf.read(input_path, always_2d=True)
    original_channels = data.shape[1]

    # Ensure stereo for processing
    if original_channels == 1:
        data = np.hstack([data, data])

    audio = data.T.astype(np.float32)  # shape: (channels, samples)

    # ------------------------------------------------------------------ #
    # 1. EQ Chain                                                          #
    # ------------------------------------------------------------------ #
    eq_settings = settings.get("eq", {})
    eq_plugins = []

    hpf_freq = eq_settings.get("highpass_hz", 80.0)
    if hpf_freq and hpf_freq > 20:
        eq_plugins.append(HighpassFilter(cutoff_frequency_hz=float(hpf_freq)))

    lpf_freq = eq_settings.get("lowpass_hz", 18000.0)
    if lpf_freq and lpf_freq < 20000:
        eq_plugins.append(LowpassFilter(cutoff_frequency_hz=float(lpf_freq)))

    low_shelf = eq_settings.get("low_shelf", {})
    if low_shelf:
        eq_plugins.append(LowShelfFilter(
            cutoff_frequency_hz=float(low_shelf.get("freq", 120)),
            gain_db=float(low_shelf.get("gain_db", 0)),
            q=float(low_shelf.get("q", 0.707)),
        ))

    high_shelf = eq_settings.get("high_shelf", {})
    if high_shelf:
        eq_plugins.append(HighShelfFilter(
            cutoff_frequency_hz=float(high_shelf.get("freq", 10000)),
            gain_db=float(high_shelf.get("gain_db", 0)),
            q=float(high_shelf.get("q", 0.707)),
        ))

    for band in eq_settings.get("peak_bands", []):
        gain_db = float(band.get("gain_db", 0))
        if abs(gain_db) > 0.1:
            eq_plugins.append(PeakFilter(
                cutoff_frequency_hz=float(band.get("freq", 1000)),
                gain_db=gain_db,
                q=float(band.get("q", 1.0)),
            ))

    if eq_plugins:
        board_eq = Pedalboard(eq_plugins)
        audio = board_eq(audio, sr)

    # ------------------------------------------------------------------ #
    # 2. Compression                                                       #
    # ------------------------------------------------------------------ #
    comp_settings = settings.get("compression", {})
    threshold_db = float(comp_settings.get("threshold_db", -18.0))
    ratio = float(comp_settings.get("ratio", 3.0))
    attack_ms = float(comp_settings.get("attack_ms", 10.0))
    release_ms = float(comp_settings.get("release_ms", 100.0))
    makeup_db = float(comp_settings.get("makeup_gain_db", 2.0))

    board_comp = Pedalboard([
        Compressor(
            threshold_db=threshold_db,
            ratio=ratio,
            attack_ms=attack_ms,
            release_ms=release_ms,
        ),
        Gain(gain_db=makeup_db),
    ])
    audio = board_comp(audio, sr)

    # ------------------------------------------------------------------ #
    # 3. Reverb (optional, wet/dry controlled)                            #
    # ------------------------------------------------------------------ #
    reverb_settings = settings.get("reverb", {})
    wet_level = float(reverb_settings.get("wet_level", 0.0))
    if wet_level > 0.01:
        board_reverb = Pedalboard([
            Reverb(
                room_size=float(reverb_settings.get("room_size", 0.3)),
                damping=float(reverb_settings.get("damping", 0.5)),
                wet_level=wet_level,
                dry_level=float(reverb_settings.get("dry_level", 0.9)),
                width=float(reverb_settings.get("width", 0.5)),
            )
        ])
        audio = board_reverb(audio, sr)

    # ------------------------------------------------------------------ #
    # 4. Stereo Width                                                      #
    # ------------------------------------------------------------------ #
    stereo_settings = settings.get("stereo", {})
    width_factor = float(stereo_settings.get("width_factor", 1.0))
    if abs(width_factor - 1.0) > 0.05 and audio.shape[0] == 2:
        mid = (audio[0] + audio[1]) * 0.5
        side = (audio[0] - audio[1]) * 0.5 * width_factor
        audio[0] = mid + side
        audio[1] = mid - side

    # ------------------------------------------------------------------ #
    # 5. LUFS Normalization                                               #
    # ------------------------------------------------------------------ #
    lufs_target = float(settings.get("lufs_target", -14.0))
    audio_out = audio.T  # back to (samples, channels)

    meter = pyln.Meter(sr)
    try:
        current_lufs = meter.integrated_loudness(audio_out)
        if current_lufs > -70 and not np.isinf(current_lufs):
            gain_needed = lufs_target - current_lufs
            gain_linear = 10 ** (gain_needed / 20.0)
            audio_out = audio_out * gain_linear
    except Exception:
        pass

    # ------------------------------------------------------------------ #
    # 6. Limiter (brick-wall at -0.3 dBFS)                               #
    # ------------------------------------------------------------------ #
    audio_for_limiter = audio_out.T.astype(np.float32)
    board_limiter = Pedalboard([
        Limiter(threshold_db=-0.3, release_ms=50.0)
    ])
    audio_for_limiter = board_limiter(audio_for_limiter, sr)
    audio_out = audio_for_limiter.T

    # Clip guard
    audio_out = np.clip(audio_out, -1.0, 1.0)

    # Write output WAV (24-bit)
    sf.write(output_path, audio_out, sr, subtype="PCM_24")

    # Measure final LUFS
    try:
        final_lufs = float(meter.integrated_loudness(audio_out))
    except Exception:
        final_lufs = lufs_target

    return {
        "output_path": output_path,
        "sample_rate": sr,
        "channels": audio_out.shape[1],
        "final_lufs": round(final_lufs, 2),
        "settings_applied": settings,
    }


if __name__ == "__main__":
    if len(sys.argv) < 4:
        print(json.dumps({"error": "Usage: process.py <input> <output_wav> <settings_json>"}))
        sys.exit(1)
    with open(sys.argv[3]) as f:
        settings = json.load(f)
    result = apply_dsp(sys.argv[1], sys.argv[2], settings)
    print(json.dumps(result))

File: server/dsp/process.ts
/**
 * process.ts — DSP processing using FFmpeg filter chains
 * Applies EQ, compression, reverb, stereo widening, and loudness normalization
 */

import * as ffmpegStatic from "ffmpeg-static";
// @ts-ignore - fluent-ffmpeg lacks type definitions
import ffmpeg from "fluent-ffmpeg";
import * as path from "path";

// Set FFmpeg path
if (ffmpegStatic && typeof ffmpegStatic === "string") {
  ffmpeg.setFfmpegPath(ffmpegStatic);
} else if (ffmpegStatic && typeof ffmpegStatic === "object" && "path" in ffmpegStatic) {
  ffmpeg.setFfmpegPath((ffmpegStatic as any).path);
}

export interface MixSettings {
  eq: {
    highpass_hz: number;
    lowpass_hz: number;
    low_shelf: { freq: number; gain_db: number; q: number };
    high_shelf: { freq: number; gain_db: number; q: number };
    peak_bands: Array<{ freq: number; gain_db: number; q: number; label: string }>;
  };
  compression: {
    threshold_db: number;
    ratio: number;
    attack_ms: number;
    release_ms: number;
    makeup_gain_db: number;
  };
  reverb: {
    room_size: number;
    damping: number;
    wet_level: number;
    dry_level: number;
    width: number;
  };
  stereo: {
    width_factor: number;
  };
  lufs_target: number;
  reasoning: Record<string, unknown>;
}

/**
 * Build FFmpeg filter chain for audio processing
 */
function buildFilterChain(settings: MixSettings): string {
  const filters: string[] = [];

  // 1. High-pass filter (remove sub-bass rumble)
  filters.push(`highpass=f=${settings.eq.highpass_hz}`);

  // 2. Low-pass filter (remove ultrasonic noise)
  filters.push(`lowpass=f=${settings.eq.lowpass_hz}`);

  // 3. Low shelf EQ
  const lowShelf = settings.eq.low_shelf;
  filters.push(`equalizer=f=${lowShelf.freq}:t=lowshelf:width_type=q:w=${lowShelf.q}:g=${lowShelf.gain_db}`);

  // 4. High shelf EQ
  const highShelf = settings.eq.high_shelf;
  filters.push(`equalizer=f=${highShelf.freq}:t=highshelf:width_type=q:w=${highShelf.q}:g=${highShelf.gain_db}`);

  // 5. Peak EQ bands
  for (const band of settings.eq.peak_bands) {
    filters.push(`equalizer=f=${band.freq}:t=peak:width_type=q:w=${band.q}:g=${band.gain_db}`);
  }

  // 6. Compressor
  const comp = settings.compression;
  filters.push(
    `acompressor=threshold=${comp.threshold_db}:ratio=${comp.ratio}:attack=${comp.attack_ms}:release=${comp.release_ms}:makeup=${comp.makeup_gain_db}`
  );

  // 7. Limiter (prevent clipping)
  filters.push(`alimiter=limit=0.99:attack=5:release=50`);

  // 8. Loudness normalization (using loudnorm filter)
  filters.push(`loudnorm=I=-14:TP=-1.5:LRA=11`);

  // 9. Stereo widening (simple mid-side processing)
  if (settings.stereo.width_factor > 1.0) {
    // Boost stereo width slightly
    filters.push(`stereotools=slevel=${settings.stereo.width_factor}`);
  }

  return filters.join(",");
}

/**
 * Apply DSP processing to audio file using FFmpeg
 */
export async function applyDSP(
  inputPath: string,
  outputPath: string,
  settings: MixSettings,
  sampleRate: number
): Promise<void> {
  return new Promise((resolve, reject) => {
    console.log(`[DSP] Applying processing chain to: ${inputPath}`);
    console.log(`[DSP] Output: ${outputPath}`);
    console.log(`[DSP] Settings:`, JSON.stringify(settings, null, 2));

    const filterChain = buildFilterChain(settings);
    console.log(`[DSP] Filter chain: ${filterChain}`);

    ffmpeg(inputPath)
      .audioFilters(filterChain)
      .audioCodec("pcm_s24le")
      .audioFrequency(sampleRate)
      .audioChannels(2)
      .on("start", (cmd: any) => {
        console.log(`[DSP] FFmpeg command: ${cmd}`);
      })
      .on("progress", (progress: any) => {
        console.log(`[DSP] Processing progress: ${progress.percent}%`);
      })
      .on("error", (err: any) => {
        console.error(`[DSP] FFmpeg error:`, err);
        reject(err);
      })
      .on("end", () => {
        console.log(`[DSP] Processing complete`);
        resolve();
      })
      .save(outputPath);
  });
}

/**
 * Convert WAV to MP3 using FFmpeg
 */
export async function convertToMP3(wavPath: string, mp3Path: string): Promise<void> {
  return new Promise((resolve, reject) => {
    console.log(`[MP3] Converting WAV to MP3: ${wavPath} -> ${mp3Path}`);

    ffmpeg(wavPath)
      .audioCodec("libmp3lame")
      .audioBitrate("320k")
      .audioFrequency(44100)
      .on("error", (err: any) => {
        console.error(`[MP3] FFmpeg error:`, err);
        reject(err);
      })
      .on("end", () => {
        console.log(`[MP3] Conversion complete`);
        resolve();
      })
      .save(mp3Path);
  });
}
